File: backend-nodejs/controllers/CouponController.js
const Coupon = require('../models/Coupon');
const CouponUsage = require('../models/CouponUsage');

class CouponController {
  async validateCoupon(req, res) {
    try {
      const { code } = req.params;
      const { userId } = req.query;
      
      const coupon = await Coupon.findOne({ where: { code } });
      
      if (!coupon) {
        return res.status(404).json({ error: 'Coupon not found' });
      }
      
      if (coupon.status !== 'ACTIVE') {
        return res.status(400).json({ error: 'Coupon not active' });
      }
      
      const usage = await CouponUsage.findOne({
        where: { userId, couponId: coupon.id }
      });
      
      if (usage) {
        return res.status(409).json({ error: 'Coupon already used' });
      }
      
      res.json(coupon);
    } catch (error) {
      res.status(500).json({ error: error.message });
    }
  }

  async getCoupon(req, res) {
    try {
      const { code } = req.params;
      const coupon = await Coupon.findOne({ where: { code } });
      
      if (!coupon) {
        return res.status(404).json({ error: 'Coupon not found' });
      }
      
      res.json(coupon);
    } catch (error) {
      res.status(500).json({ error: error.message });
    }
  }
}

module.exports = new CouponController();

File: genaibackend/src/main/java/com/clothingstore/controller/CouponController.java
package com.clothingstore.controller;

import com.clothingstore.entity.Coupon;
import com.clothingstore.entity.CouponUsage;
import com.clothingstore.repository.CouponRepository;
import com.clothingstore.repository.CouponUsageRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.Optional;

@RestController
@RequestMapping("/api/products/coupons")
@CrossOrigin(origins = "*")
public class CouponController {
    
    @Autowired
    private CouponRepository couponRepository;
    
    @Autowired
    private CouponUsageRepository couponUsageRepository;
    
    @GetMapping("/validate/{code}")
    public ResponseEntity<Coupon> validateCoupon(@PathVariable String code, @RequestParam Long userId) {
        Optional<Coupon> couponOpt = couponRepository.findByCode(code);
        
        if (!couponOpt.isPresent()) {
            return ResponseEntity.notFound().build();
        }
        
        Coupon coupon = couponOpt.get();
        
        // Check if coupon is active
        if (coupon.getStatus() != Coupon.CouponStatus.ACTIVE) {
            return ResponseEntity.badRequest().build();
        }
        
        // Check if user has already used this coupon
        Optional<CouponUsage> usageOpt = couponUsageRepository.findByUserIdAndCouponId(userId, coupon.getId());
        if (usageOpt.isPresent()) {
            return ResponseEntity.status(409).build(); // Conflict - already used
        }
        
        return ResponseEntity.ok(coupon);
    }
    
    @GetMapping("/{code}")
    public ResponseEntity<Coupon> getCoupon(@PathVariable String code) {
        Optional<Coupon> coupon = couponRepository.findByCode(code);
        return coupon.map(ResponseEntity::ok).orElse(ResponseEntity.notFound().build());
    }
}